src/main/java/org/lmdbjava/Cursor.java

```diff
@@ -204,11 +211,20 @@ public void put(final T key, final T val, final PutFlags... op) {
     }
     kv.keyIn(key);
     kv.valIn(val);
-    final int flags = mask(op);
-    checkRc(LIB.mdb_cursor_put(ptrCursor, kv.pointerKey(), kv.pointerVal(),
-                               flags));
-    kv.keyOut();
-    kv.valOut();
+    final int mask = mask(op);
+    final int rc = LIB.mdb_cursor_put(ptrCursor, kv.pointerKey(), kv.pointerVal(),
+                               mask);
+    if (rc == MDB_KEYEXIST) {
+      if (isSet(mask, MDB_NOOVERWRITE)) {
+        kv.valOut(); // marked as in,out in LMDB C docs
+      } else if (!isSet(mask, MDB_NODUPDATA)) {
+        checkRc(rc);
+      }
+      return false;
+    } else {
+      checkRc(rc);
+      return true;
+    }
```

Comment:
Please remove the final `else`, as it's easier to read when the `MDB_KEYEXIST` logic is in its own distinct `if` block.